orderedmap.go

```diff
@@ -109,6 +113,67 @@ func (om *OrderedMap[K, V]) Newest() *Pair[K, V] {
 	return listElementToPair(om.list.Back())
 }
 
+// MarshalJSON implements the json.Marshaler interface.
+func (om *OrderedMap[K, V]) MarshalJSON() ([]byte, error) {
+	buf := bytes.Buffer{}
+	buf.WriteString("{")
+
+	i := 0
+	for pair := om.Oldest(); pair != nil; pair = pair.Next() {
+		if i > 0 {
+			buf.WriteString(",")
+		}
+
+		var marshaledKey string
+		switch key := any(pair.Key).(type) {
+		case string:
+			marshaledKey = key
+		case encoding.TextMarshaler:
+			marshaledKeyBytes, err := key.MarshalText()
+			if err != nil {
+				return nil, err
+			}
+			marshaledKey = string(marshaledKeyBytes)
+		case int:
+			marshaledKey = strconv.FormatInt(int64(key), 10)
+		case int8:
+			marshaledKey = strconv.FormatInt(int64(key), 10)
+		case int16:
+			marshaledKey = strconv.FormatInt(int64(key), 10)
+		case int32:
+			marshaledKey = strconv.FormatInt(int64(key), 10)
+		case int64:
+			marshaledKey = strconv.FormatInt(int64(key), 10)
+		case uint:
+			marshaledKey = strconv.FormatInt(int64(key), 10)
+		case uint8:
+			marshaledKey = strconv.FormatInt(int64(key), 10)
+		case uint16:
+			marshaledKey = strconv.FormatInt(int64(key), 10)
+		case uint32:
+			marshaledKey = strconv.FormatInt(int64(key), 10)
+		case uint64:
+			marshaledKey = strconv.FormatInt(int64(key), 10)
+		default:
+			return nil, fmt.Errorf("unsupported key type: %T", key)
+		}
+		marshaledKey = `"` + marshaledKey + `"`
+
+		value, err := json.Marshal(pair.Value)
+		if err != nil {
+			return nil, err
+		}
+
+		buf.WriteString(marshaledKey + ":")
+		buf.Write(value)
+		i++
+	}
+
+	buf.WriteString("}")
```

Comment:
https://pkg.go.dev/bytes#Buffer.WriteRune (and avoid concatanation of small-ish strings)

Reply:
I don't see a difference in perf (`WriteString` vs `WriteRune`):

```
➜  go-ordered-map git:(marshal-json) go test -bench=. -benchmem -benchtime=10s
goos: darwin
goarch: arm64
pkg: github.com/wk8/go-ordered-map/v2
BenchmarkMarshalJSON-8           8609314              1382 ns/op             352 B/op         11 allocs/op
PASS
ok      github.com/wk8/go-ordered-map/v2        13.655s
➜  go-ordered-map git:(marshal-json) go test -bench=. -benchmem -benchtime=10s
goos: darwin
goarch: arm64
pkg: github.com/wk8/go-ordered-map/v2
BenchmarkMarshalJSON-8           8678156              1383 ns/op             352 B/op         11 allocs/op
PASS
ok      github.com/wk8/go-ordered-map/v2        13.693s
```